From 608b812b4ca1e3ebf283cd03ab856682062d179e Mon Sep 17 00:00:00 2001
From: Eliad Peller <eliad@wizery.com>
Date: Thu, 7 Apr 2011 15:08:17 +0300
Subject: [PATCH 28/75] wl12xx: croc only after interface is up

This way we will remain-on-channel during the EAPOLs exchange.

In the future we should probably use wpa_supplicant to configure
these roc/croc.
---
 drivers/net/wireless/wl12xx/main.c |   21 ++++++++++++++-------
 1 files changed, 14 insertions(+), 7 deletions(-)

diff --git a/drivers/net/wireless/wl12xx/main.c b/drivers/net/wireless/wl12xx/main.c
index d6c08db..569d74f 100644
--- a/drivers/net/wireless/wl12xx/main.c
+++ b/drivers/net/wireless/wl12xx/main.c
@@ -466,6 +466,7 @@ static int wl1271_dev_notify(struct notifier_block *me, unsigned long what,
 	if ((dev->operstate == IF_OPER_UP) &&
 	    !test_and_set_bit(WL1271_FLAG_STA_STATE_SENT, &wl->flags)) {
 		wl1271_cmd_set_peer_state(wl);
+		wl1271_croc(wl);
 		wl1271_info("Association completed.");
 	}
 
@@ -3157,7 +3158,9 @@ static void wl1271_bss_info_changed_sta(struct wl1271 *wl,
 			bool was_assoc =
 			    !!test_and_clear_bit(WL1271_FLAG_STA_ASSOCIATED,
 						 &wl->flags);
-			clear_bit(WL1271_FLAG_STA_STATE_SENT, &wl->flags);
+			bool was_ifup =
+			    !!test_and_clear_bit(WL1271_FLAG_STA_STATE_SENT,
+						 &wl->flags);
 			wl->aid = 0;
 
 			/* free probe-request template */
@@ -3184,6 +3187,15 @@ static void wl1271_bss_info_changed_sta(struct wl1271 *wl,
 
 			/* restore the bssid filter and go to dummy bssid */
 			if (was_assoc) {
+				/*
+				 * we might have to disable roc, if there was
+				 * no IF_OPER_UP notification.
+				 */
+				if (!was_ifup) {
+					ret = wl1271_croc(wl);
+					if (ret < 0)
+						goto out;
+				}
 				wl1271_unjoin(wl);
 				wl1271_roc(wl);
 			}
@@ -3223,12 +3235,7 @@ static void wl1271_bss_info_changed_sta(struct wl1271 *wl,
 	}
 
 	if (do_join) {
-		/* disable ROC before joining */
-		if (test_bit(WL1271_FLAG_ROC, &wl->flags)) {
-			ret = wl1271_croc(wl);
-			if (ret < 0)
-				goto out;
-		}
+		/* don't cancel ROC (on dev role) until interface is up */
 		ret = wl1271_join(wl, set_assoc);
 		if (ret < 0) {
 			wl1271_warning("cmd join failed %d", ret);
-- 
1.7.0.4

